include/seastar/net/api.hh

```diff
@@ -129,6 +129,22 @@ public:
     void close();
 };
 
+class input_buffer_factory {
+public:
+  using buffer_t = temporary_buffer<char>;
+
+  virtual ~input_buffer_factory() = default;
+  /// Provide a rx buffer. Implementation is responsible for determining its size
+  /// and memory. This is useful when a network stack implementation does not put
+  /// extra requirements on these factors. The POSIX stack is the example here.
+  /// \param allocator Memory allocator \c connected_socket implementation prefers.
+  /// Maybe nullptr.
+  virtual buffer_t create(compat::polymorphic_allocator<char>* allocator) = 0;
+
+  // Give back to the factory unused part of a buffer obtained from it
+  virtual void return_unused(buffer_t&& buf) = 0;
```

Comment:
Curious about how to make use of the unused tail of the rx buffer, feed back to the data_souce later?

Reply:
Yup, that's the idea. For huge chunks even the POSIX stack tends to fragment the output. This is likely because of the `readable(fd)` triggering just when *something* becomes available to read:
```cpp
future<size_t>
reactor::read_some(pollable_fd_state& fd, void* buffer, size_t len) {
    return readable(fd).then([this, &fd, buffer, len] () mutable {
        auto r = fd.fd.read(buffer, len);
        // ...
}
```